### PR TITLE
[WIP] GitLab CI

### PR DESCRIPTION
**Pull request by D4N**
As discussed during the team meeting, I have prepared a config for GitLab CI and a build script that runs a build & all tests for all combinations of the following options:
- shared/static library
- release/debug build
- compilation with gcc or clang

The GitLab build covers the (imho) most widely used Linux distributions (I have not included derived distros that stay close to their "parent" like Linux Mint or Manjaro):
- Fedora
- Archlinux
- Ubuntu
- Debian
- CentOS
- OpenSUSE

During experimentation I have already uncovered some issues:
1. The safe_op.hpp file was fully specializing template functions in a header which is an ODR violation. The template functions are now declared `inline`.
2. Some compiler flags for gcc & clang were erroneously added for compiler versions that do not support them (for clang I was unfortunately forced to guess, since I couldn't find a changelog of the added flags for past versions).

Still left to do:
- [x] clean up the git history, its a bloody mess due to a lot of experimenting
- [x] make GitLab actually run the pipeline on pull requests (only on PRs from branches on exiv2, see: [GitLab EE issue 5667](https://gitlab.com/gitlab-org/gitlab-ee/issues/5667))
- [ ] ~~run the same build script on Travis (maybe even appveyor) too?~~

Unfortunately there is currently a big catch: GitLab does not run the pipeline for pull requests from forks but only for branches on Exiv2/exiv2. Unfortunately there is nothing we can do about that :(
I would however still advocate to include the GitLab config as most of the contributions are from team members who can create new branches so the CI situation will improve, albeit not as much as I hoped it will.

### CODE REVIEW

**Comment by D4N** [comment]
Ok, the gitlab CI is now showing up on github!

File: config/compilerFlags.cmake [hunk 1]
```diff
@@ -72,6 +75,18 @@ if ( MINGW OR UNIX ) # MINGW, Linux, APPLE, CYGWIN
                     #" -Wconversion"
                     #" -Wold-style-cast"
                 )
+                # -Wdouble-promotion flag is not available in clang 3.4.2
```
Comment (piponazo) [inline config/compilerFlags.cmake]: [question] Just out of curiosity. Which distribution is using such old version of clang ? :astonished:

**Review by piponazo — APPROVED** [review]
Wow man, this looks amazing! 

As I told you via Slack, I think it is not a big deal the fact of not having GitLab CI checks for PRs coming from forks. Now we know that it is building for all the different distributions you added here. The moment something get red on master, we can create a PR from the main repository to fix that situation.

Thanks a lot for this contribution.

By the way, maybe you can add something in the main README.md to show a status image badge.

**Comment by D4N** [comment]
I'll add the badge, that's a good idea.

What do you think about using the same build script on travis (since we only run release builds + dynamic library there)?

**Comment by piponazo** [comment]
> What do you think about using the same build script on travis (since we only run release builds + dynamic library there)?

I have double feelings about that. On the one hand, the more testing we have, the better. You can see that I actually tried several combinations with travis in the past (commented lines in the .travis.yml file). However, on the other hand, having such amount of combinations was making much slower the CI feedback.

Now that we will have Gitlab CI scripts, with all the combinations that you have introduced there, I think that we have a nice and balanced safety net:

- TravisCI and AppVeyor: Quick checks for Linux, Mac and Windows. External dependencies handled by conan. Standard configuration in all the cases: Usage of all the 3rd party dependencies + Release mode.
- Gitlab: A larger amount of combinations for the main linux distributions. Using system packages for the external dependencies. It will take more time to run those CI checks, but we will run them only on the master branch (not for every PR that other contributors might create).

I would stay with the current scripts from travisCI for the moment. If we find the need of using the same scripts there, we can always do it in the future.

**Comment by D4N** [comment]
Sure, makes sense, lets keep travis the way it is for now.

Just a last question, I have created a new folder `ci/` where I put the scripts for gitlab. Should we merge this folder with `.travis/`? So that we don't end up with a bunch of folders in the root directory.

**Comment by piponazo** [comment]
That's fine for me. I think it makes sense to move anything related to CI to a common folder (as much as possible).

**Pull request merged**